engines/dfc_provider/app/services/dfc_request.rb

```diff
@@ -4,31 +4,66 @@
 require "private_address_check/tcpsocket_ext"
 
 # Request a JSON document from a DFC API with authentication.
+#
+# All DFC API interactions are authenticated via OIDC tokens. If the user's
+# access token is expired, we try to get a new one with the user's refresh
+# token.
 class DfcRequest
   def initialize(user)
     @user = user
   end
 
   def get(url)
+    response = request(url)
+
+    return response.body if response.status == 200
+
+    return "" if @user.oidc_account.updated_at > 15.minutes.ago
+
+    refresh_access_token!
+
+    response = request(url)
+    response.body
```

Comment:
Something didn't sit right with me: the main success path isn't very obvious (at least to me). 
I think it's because we return early on success, and I would generally prefer to see returns near the end. It means the bulk of this method is handling the "exception".

So this is just my opinion, but I would suggest that the retry logic be moved to another method. Then it could be something like:
```
    response = request(url)

    response = refresh_and_retry_request(url) if response.status != 200

    response.body
```

Reply:
I wasn't happy with that either but I couldn't think of a really good solution. I like your pattern though, that's a nice flow. I'll do that in a follow up, I think.